### ISSUE TITLE
[HOLD][$1000] Unable to edit the message when multiple messages are received 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18805]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app in android and login with user A
2. Open report of user B
3. edit any message, do not save it, if needed, you can exit the report
4. From any other device, login with user B
5. send multiple messages to user A until edited message leaves the screeen on user A device
6. On user A device, in user B report, click on edit message compose box to edit message and observe that app does not open the keyboard, does not display edit message compose box in middle and glitches out

## Expected Result: 
App should allow to edit the message
## Actual Result: 
App does not open edit message keyboard and does not allow us to edit the message if multiple messages are received while editing the message
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.13.0 
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/0e7bc6b8-f7a0-4115-b628-4b0e97678fea

https://github.com/Expensify/App/assets/43996225/d4e5bf51-565d-494d-82af-dc7a7eb97c9f

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1683777332636589

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01abb91265e8834d00</li>
        <li>Upwork Job ID: 1656952977054523392</li>
        <li>Last Price Increase: 2023-05-12</li>
    </ul>
</details>

**Comment by jeet-dhandha** [comment]
Few questions : 
- Why do we keep draft message id ?
- When closing a chat why aren't they getting removed ?
- Why are we allowing multiple edits at the same time for the user ?

**Comment by mountiny** [comment]
> Why are we allowing multiple edits at the same time for the user ?

This is just a design decision, its on purpose

I think this would be great issue for Callstack so asking them to take this one cc @marcochavezf

**Comment by Santhosh-Sellavel** [comment]
Waiting for proposals here.

**Comment by gabrialva** [comment]
where can I find the android code base? I think I can fix this issue, but I am gonna check the source code on my local

**Comment by Piotrfj** [comment]
Hi I'm Piotr from Callstack - expert contributor group - will start investigation in this area

**Comment by flaviadefaria** [comment]
A contributor was just assigned Melvin!

**Comment by flaviadefaria** [comment]
I'm OoO for 8 days so re-adding the BUG label, but keeping myself assigned. I'll be back at work on May 30th.

**Comment by twisterdotcom** [comment]
Any update on this @Piotrfj?

**Comment by Piotrfj** [comment]
Update: last week I had some setup issues, but i was finally able to reproduce the bug and was investigating what is the source of the issues, hope to find it soon.

**Comment by Piotrfj** [comment]
Update: Found probable reason of the bug, need to investigate it further

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj  Any update?

**Comment by Piotrfj** [comment]
Yes, was working to the late hour and was trying to solve the issue, but fixing the probable source did not solve it completely, today we have got workshops so I will jump back to it at monday

**Comment by Santhosh-Sellavel** [comment]
Any update @Piotrfj

**Comment by Piotrfj** [comment]
I have an draft of the proposal ready, will review it internally tomorrow morning along with proposition of another probably unrelated improvement.

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj Can you follow our proposal [template](https://raw.githubusercontent.com/Expensify/App/main/contributingGuides/PROPOSAL_TEMPLATE.md) as it is more readable

**Comment by twisterdotcom** [comment]
Unassigning now @flaviadefaria is returned.

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj Can you explain elaborately with changes and include more information on the root cause of which block of code causes, make sure to follow our proposal [template](https://raw.githubusercontent.com/Expensify/App/main/contributingGuides/PROPOSAL_TEMPLATE.md)

cc: @marcochavezf

**Comment by Piotrfj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The problem we are trying to solve is edit input self closing after someone else spammed the chat.

### What is the root cause of that problem?
The index property in ReportActionItem is not being checked in shouldComponentUpdate lifecycle method therefore the edited item is not being updated when new items (messages) are added below and in result we are ending up with 2 items that are represented with the same index, thats why the focus and scrolling mechanism is glitching.

### What changes do you think we should make in order to solve the problem?
Check the index along with existing rules to shouldComponentUpdate method.

The block that causes the issue:
`shouldComponentUpdate(nextProps, nextState) {
        return (
            this.props.displayAsGroup !== nextProps.displayAsGroup ||
            this.props.draftMessage !== nextProps.draftMessage ||
            this.props.isMostRecentIOUReportAction !== nextProps.isMostRecentIOUReportAction ||
            this.props.hasOutstandingIOU !== nextProps.hasOutstandingIOU ||
            this.props.shouldDisplayNewMarker !== nextProps.shouldDisplayNewMarker ||
            !_.isEqual(this.props.action, nextProps.action) ||
            this.state.isContextMenuActive !== nextState.isContextMenuActive ||
            lodashGet(this.props.report, 'statusNum') !== lodashGet(nextProps.report, 'statusNum') ||
            lodashGet(this.props.report, 'stateNum') !== lodashGet(nextProps.report, 'stateNum') ||
            this.props.translate !== nextProps.translate
        );
    }`

### What alternative solutions did you explore?
Appearing messages are still pushing our edit input up, so this is probably something we would like to fix by holding the new messages under the input during edit mode (block the scrolling mechanism in place).

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj 
Which file the block exists in?

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel In ReportActionItem, source path: src/pages/home/report/ReportActionItem.js

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj 
> the index property in ReportActionItem is not being checked in shouldComponentUpdate lifecycle method therefore the edited item is not being updated when new items (messages) are added below and in result we are ending up with 2 items that are represented with the same index, thats why the focus and scrolling mechanism is glitching.

Sorry for the delay if that's the case, I'm wondering why it's happening on Android alone.

**Comment by Santhosh-Sellavel** [comment]
Waiting for @Piotrfj

**Comment by Piotrfj** [comment]
Update: The component where the bug was present was refactored in the meantime, currently investigating the solution and the difference between platforms. Will provide some more info later today

**Comment by Santhosh-Sellavel** [comment]
@Expensify/design thoughts? Here is how Slack does for mobile for quick reference.
![Screenshot_20230616_002802](https://github.com/Expensify/App/assets/85645967/304ee95e-e972-4d52-8f38-2d5ac9c617cb)

cc: @marcochavezf

**Comment by shawnborton** [comment]
I'm not exactly sure what you are asking me - mind rephrasing with more details?

**Comment by Piotrfj** [comment]
@shawnborton I'm not sure if you are asking me to rephrase the proposed solution or @Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
@shawnborton Sorry for the delay, I missed your comment. For this issue, @Piotrfj proposes to address the issue in a way that Slack does it.  And [here](https://github.com/Expensify/App/issues/18805#issuecomment-1593578815) I've added a screenshot of how it looks on Slack.

#### @Piotrfj's proposal
> I would like to propose a little refactor that would fix the current issue and on top it would make editing a lot more pleasant. I could use the message composer, that is stuck to the bottom to edit the message (like it is done for example in the mobile slack version) instead of changing the item to edit input. This way we would avoid the situation where the edited message is going up when new messages are appearing at the bottom (which causes the keyboard to disappear), and the situation where the composer and edited input are visible at the same time (like in the screen below). Other solutions are mostly just workarounds.

The question for the design is, should we consider this?

**Comment by shawnborton** [comment]
Hmm we had a lot of back and forth about this in Slack if I recall. I would bring this discussion to the open source room for more opinions before we decide anything here.

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj Please initiate the thread on #expensify-open-source with your proposal with any additional context which would help us make this decision. So we can get a consensus for moving this forward thanks!

**Comment by jeet-dhandha** [comment]
Like discussed above i tried implementing it like slack : 
(there are few scenarios to handle but over all its working).

https://github.com/Expensify/App/assets/78416198/0244ebbb-bc59-4c75-b485-eb64f120f97d

**Comment by Santhosh-Sellavel** [comment]
@jeet-dhandha We aren't looking for any external  contributors proposal, this is already being worked by @Piotrfj

**Comment by jeet-dhandha** [comment]
But wasn't this an external Bug.

**Comment by Santhosh-Sellavel** [comment]
It was, now it's being worked by @Piotrfj.

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj bump
> @Piotrfj Please initiate the thread on #expensify-open-source with your proposal with any additional context which would help us make this decision. So we can get a consensus for moving this forward thanks!

Have you started the thread if not then please start and if done. Please leave the link here or tag me in the thread so I can follow up thanks!

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel just tagged you

**Comment by flaviadefaria** [comment]
The PR is being worked on.

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel One of the thing that will work in this case is disabling the animation of the scroll when clicking the edited item. Ill provide a video example of how this would work.

Im also thinking of blocking automatic scroll when new items are appearing, so the edited message will not be pushed to the top when someone writes something in the chat during editing. Working on this solution right now

**Comment by Piotrfj** [comment]
This is how it work without the animation:


https://github.com/Expensify/App/assets/41384305/26dd195b-dc99-4d1e-90e1-92942d2f9110

**Comment by Piotrfj** [comment]
Thats how it looks with animation:

Uploading Kapture 2023-06-30 at 17.04.16.mp4…

**Comment by Piotrfj** [comment]
https://github.com/Expensify/App/assets/41384305/426ea9c7-8d94-4ae3-a5f0-d56e69b52146

**Comment by flaviadefaria** [comment]
@shawnborton @marcochavezf what are your thoughts on the above?

**Comment by shawnborton** [comment]
I don't think we should change the mechanism for how we edit messages. There was a ton of context about how that should work. 

If anything, maybe this particular bug would benefit from comment linking? Like perhaps there is a way to lock the screen position when editing messages, even if new ones are rolling in below. Thoughts @JmillsExpensify @trjExpensify ?

But I would not change the way we edit messages to accommodate this edge case right now.

**Comment by trjExpensify** [comment]
Yeah, this is an interesting one. I tend to agree that it's jarring to be in the middle of an edit in a buoyant room and other people adding messages to the chat scrolls your edit composer off the screen. So I agree with the notion of locking the screen position there.

**Comment by flaviadefaria** [comment]
Cool, I like that and I think it's what Slack does too - it adds a line showing the number of unread messages you have below but locks the screen position you're in.

**Comment by jeet-dhandha** [comment]
https://github.com/Expensify/App/issues/22368 - This should also be handled here.

**Comment by jeet-dhandha** [comment]
https://github.com/Expensify/App/issues/15596 - this should also be handled

**Comment by jeet-dhandha** [comment]
https://github.com/Expensify/App/issues/16843#issuecomment-1494366248 This one too

**Comment by flaviadefaria** [comment]
@Piotrfj how's the PR going?

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj What's the status here?

**Comment by Piotrfj** [comment]
Working on the issue, will come with proposal today or tomorrow

**Comment by flaviadefaria** [comment]
Nice, looking forward to it!

**Comment by Piotrfj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The problem we are trying to solve is edit input self closing after someone else spammed the chat.

### What is the root cause of that problem?
When someone else writes too many messages in the chat, the edit input is pushed out of the visible area causing the input to blur and the keyboard to hide, after focusing input again when it is high on the screen, the scroll animation is too slow comparing to keyboard and the keyboard is pushing the content up causing the input to disappear again from visible area.

### What changes do you think we should make in order to solve the problem?
Remove scroll animation and block automatic scroll when editing the message, so if someone writes message it will not push our input up.

Core of the solution for blocking messages is to put this code
`maintainVisibleContentPosition={{
  autoscrollToTopThreshold: 10,
  minIndexForVisible: 1,
}}`

Inside `InvertedFlatList` inside `ReportActionsList.js` file


### What alternative solutions did you explore?
Refactoring of editing flow, but it was rejected before

**Comment by flaviadefaria** [comment]
@Santhosh-Sellavel what are your thoughts on @Piotrfj's new proposal?

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj So far Okay sounds good. As this is a sensitive component can you write a full solution so we can get input & feedback from others too before proceeding to the PR? 

Thanks!

**Comment by Piotrfj** [comment]
Today I'm OOO, will provide a written solution tomorrow

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj any update?

**Comment by flaviadefaria** [comment]
@Piotrfj do you have an update? This is starting to take longer than it should so would appreciate it if you could focus on it. Thanks!

**Comment by Piotrfj** [comment]
Full solution:

In `ReportActionItemMessageEdit.js` file change in this line
[App/src/pages/home/report/ReportActionItemMessageEdit.js]
https://github.com/Expensify/App/blob/33ed1672b7292725081ef73caf80667a77075357/src/pages/home/report/ReportActionItemMessageEdit.js#L324C11-L324C11

`animated: true` to `animated: false`

This will remove scroll animation, so when user will focus edited item it will not self close. The animation here is the problem because it takes some time and during this time keyboard is able to push the edited item outside of visible area, which is causing the keyboard to hide.

In `ReportActionsList.js` file in `InvertedFlatList ` component add this line of code as property:

`maintainVisibleContentPosition={{ autoscrollToTopThreshold: 10, minIndexForVisible: 1, }}`

This will hold the position of the list while new items are appearing (like in the situation when someone writes a new message in the chat).

**Comment by JmillsExpensify** [comment]
> Yeah, this is an interesting one. I tend to agree that it's jarring to be in the middle of an edit in a buoyant room and other people adding messages to the chat scrolls your edit composer off the screen. So I agree with the notion of locking the screen position there.

Late though I also agree with this.

**Comment by Santhosh-Sellavel** [comment]
> In `ReportActionsList.js` file in `InvertedFlatList ` component add this line of code as property:
> 
> `maintainVisibleContentPosition={{ autoscrollToTopThreshold: 10, minIndexForVisible: 1, }}`

Wouldn't this cause a regression, where we are not editing a message and we are still stuck seeing old messages?

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel after checking indeed it would cause some problems.
I was working today on the solution for that and my proposal is to create a new ONYX flag (for example `SHOULD_FREEZE_CHAT_SCROLL`) and use that as condition of the argument that is freezing the auto scroll. It would be set in onFocus event in `ReportActionItemMessageEdit` component and unset in onBlur event in the same component.
Tested it and works fine. What do you think?

**Comment by Santhosh-Sellavel** [comment]
IMO Onyx is a bit too much for this, let me know if you have any other alternate solutions, thanks?

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel I could use useContext instead of ONYX, will work the same. What do you think?

**Comment by flaviadefaria** [comment]
@Santhosh-Sellavel thoughts on @Piotrfj's question above?

**Comment by Santhosh-Sellavel** [comment]
@Piotrfj 
Sorry, I completed miss this one, `useContext` looks good to me

Can I get a full solution on how you are planning to use it?

Considering the fact we can open multiple edits as well?

**Comment by marcochavezf** [comment]
Not overdue

**Comment by Santhosh-Sellavel** [comment]
Bump @Piotrfj

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel I was working on a solution yesterday, going to provide a written solution today soon.

**Comment by Piotrfj** [comment]
@Santhosh-Sellavel All code differences are visible in this pr https://github.com/Expensify/App/pull/24154.

Explanation in couple words:
I've created a context in new file and used it as a way to pass state setter (a simple boolean that determines whether we should freeze the scroll) down to children.

The setter is only being used in `ReportActionItemMessageEdit` component.
I'm setting it to true in onFocus action and to false in onBlur and in deleteDraft (because I have noticed that onBlur does not fire when we are closing the edited message) actions. 
I haven't found any issues with multiple edited messages, onFocus fires after onBlur, so boolean is on the right state.

Waiting for feedback about the solution.

**Comment by flaviadefaria** [comment]
@Piotrfj @Santhosh-Sellavel when do we expect to have a PR for this? It's taking a really long time to get this over the finish line after having the issue assigned.

**Comment by Piotrfj** [comment]
@flaviadefaria Draft PR is ready and linked, waiting for @Santhosh-Sellavel feedback.

**Comment by marcochavezf** [comment]
> @flaviadefaria Draft PR is ready and linked, waiting for @Santhosh-Sellavel feedback.

Not overdue, waiting for @Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
Sorry, I missed this since we already have a draft PR and [this](https://github.com/Expensify/App/issues/18805#issuecomment-1665472371) seems fine to me thanks.

We can just start reviewing and continue testing over there. @Piotrfj Can you make PR ready for review?

cc: @marcochavezf

**Comment by flaviadefaria** [comment]
I'm heading OoO today for a week so reassigning this until I'm back.

**Comment by Piotrfj** [comment]
The PR is being reviewed right now.

**Comment by dylanexpensify** [comment]
Reviewing tomorrow! Any update on review @Piotrfj

**Comment by dylanexpensify** [comment]
@Piotrfj can we get an update

**Comment by dylanexpensify** [comment]
Perfect, ty @TomaszFrackowiak <3

**Comment by dylanexpensify** [comment]
updating today!

**Comment by dylanexpensify** [comment]
Looking for an update today @Piotrfj 🙇‍♂️

**Comment by flaviadefaria** [comment]
Hey @dylan I'm back so can take over again. Thanks for you help! 
@Piotrfj do you have an update?

**Comment by Santhosh-Sellavel** [comment]
PR Review in progress

**Comment by dylanexpensify** [comment]
Amazing! Thanks Flavia! 🙇‍♂️

**Comment by flaviadefaria** [comment]
PR is in review

**Comment by flaviadefaria** [comment]
What is the latest here @Santhosh-Sellavel? It seems like the PR has been in review for over 2 weeks now.

**Comment by situchan** [comment]
@Santhosh-Sellavel is OOO this week - https://expensify.slack.com/archives/C02NK2DQWUX/p1694199491598169
As I [commented](https://github.com/Expensify/App/pull/24154#issuecomment-1712891903) in PR, I can continue C+ review on this issue if needed.

**Comment by flaviadefaria** [comment]
Sorry I missed your comment @situchan. Since tomorrow is Friday let's wait until @Santhosh-Sellavel is back and if he can't prioritize this then you can take over.

**Comment by flaviadefaria** [comment]
@abekkala  I'm going OoO for a week so assigning this to you so that you can move it forward in the meantime. I'll take it back from you once I'm online again. Thanks!

**Comment by abekkala** [comment]
@flaviadefaria this issue is quite old. (May 😬) When going ooo and reassigning for a new person to "move it forward" can you provide a summary so the new person better understand what state the issue is in and what needs to happen next.

Having to manually review the entire GH takes some time and a summary is helpful to get caught up to speed quickly and

**Comment by abekkala** [comment]
@Piotrfj's hired for fix 

Looks like we've been waiting on @Santhosh-Sellavel to review [this PR](https://github.com/Expensify/App/pull/24154#issuecomment-1712891903) for quite some time. 
@situchan - can you let me know if you've jumped in to continue the review or if @Santhosh-Sellavel will be doing this?

**Comment by Santhosh-Sellavel** [comment]
@situchan PR does not have any progress, so I keep this one assigned to myself, thanks!

@Piotrfj Can you please address the comments and resolve conflicts?

**Comment by marcochavezf** [comment]
Not overdue, PR is in review. Assigning myself as review to track it

**Comment by marcochavezf** [comment]
Linking [PR](https://github.com/Expensify/App/pull/24154) here

**Comment by marcochavezf** [comment]
PR still in review

**Comment by flaviadefaria** [comment]
Thanks @abekkala I'm back from OoO and taking this back from you now :)

**Comment by flaviadefaria** [comment]
@Piotrfj have you been able to address the comments in the PR?

**Comment by flaviadefaria** [comment]
@Piotrfj can you please give an update?

**Comment by Piotrfj** [comment]
@flaviadefaria I've been trying to reproduce the issue and I could not been able to do it, but then I saw that it only occurs in safari. So right now that I've reproduced it I'm working on the solution

**Comment by flaviadefaria** [comment]
Thanks! Do you think we can get this done by next week?

**Comment by Piotrfj** [comment]
@flaviadefaria Pretty sure I will finish it by the end of this week. I'm on a good way to complete it.

**Comment by flaviadefaria** [comment]
@anmurali I'm officially no longer in the BugZero team so reassigning this as it will not close until tomorrow.

**Comment by marcochavezf** [comment]
[PR](https://github.com/Expensify/App/pull/24154) still in review

**Comment by marcochavezf** [comment]
Hi @Piotrfj, friendly bump for the linked [PR](https://github.com/Expensify/App/pull/24154)

**Comment by marcochavezf** [comment]
Putting this issue on hold per this comment https://github.com/Expensify/App/pull/24154#issuecomment-1781083127

**Comment by Santhosh-Sellavel** [comment]
@anmurali Can we reopen this please?
